src/ck-core/cksyncbarrier.h

```diff
@@ -0,0 +1,83 @@
+#ifndef CKSYNCBARRIER_H
+#define CKSYNCBARRIER_H
+
+#include "CkSyncBarrier.decl.h"
+#include "lbdb.h"
+
+extern CkGroupID _syncBarrier;
+
+class CkSyncBarrierInit : public Chare
+{
+public:
+  CkSyncBarrierInit(CkArgMsg*);
+  CkSyncBarrierInit(CkMigrateMessage* m) : Chare(m) {}
+};
+
+CkpvExtern(bool, cksyncbarrierInited);
+
+class CkSyncBarrier : public CBase_CkSyncBarrier
+{
+public:
+  CkSyncBarrier()
+  {
+    CkpvAccess(cksyncbarrierInited) = true;
+    cur_refcount = 1;
+    iter_no = -1;
+    propagated_atsync_step = 0;
+    at_count = 0;
+    on = false;
+    startedAtSync = false;
+    rank0pe = CkMyRank() == 0;
+    reset();
+  };
+  ~CkSyncBarrier(){};
+
+  inline static CkSyncBarrier* Object()
+  {
+    return CkpvAccess(cksyncbarrierInited) ? (CkSyncBarrier*)CkLocalBranch(_syncBarrier)
+                                           : nullptr;
+  }
+
+  void propagate_atsync();
+  void recvLbStart(int lb_step, int sourcenode, int pe);
+
+  LDBarrierClient AddClient(Chare* chare, std::function<void()> fn);
+  void RemoveClient(LDBarrierClient h);
+  LDBarrierReceiver AddReceiver(std::function<void()> fn);
+  void RemoveReceiver(LDBarrierReceiver h);
+  void TurnOnReceiver(LDBarrierReceiver h);
+  void TurnOffReceiver(LDBarrierReceiver h);
+  void AtBarrier(LDBarrierClient _n_c, bool flood_atsync = false);
+  void DecreaseBarrier(int c);
+  void TurnOn()
+  {
+    on = true;
+    CheckBarrier();
+  };
+  void TurnOff() { on = false; };
+  void reset();
+
+  void CallReceivers(void);
+  void CheckBarrier(bool flood_atsync = false);
+  void ResumeClients(void);
+
+private:
+  std::list<LBClient*> clients;
+  std::list<LBReceiver*> receivers;
+
+  std::list<int> local_pes_to_notify;
```

Comment:
Why a list versus a vector?

Reply:
For `clients` (and `receivers` to some degree), because of random deletion (e.g. when the load balancer migrates AtSync elements away from the PE, the clients are removed and then readded on the receive side). Really, this should be changed to use something similar to what I implemented for the LBDatabase (sort of a linked list embedded inside a vector that keeps track of empty slots and avoids compaction when items are removed from the middle). I'm planning on making that implementation more generic, moving it into `util/` and then using it here.

Semantically, this is actually a set, but I think we can get better performance via a vector.

I've changed what was `local_pes_to_modify` to be a `vector<bool>` now.